src/app/components/hero-list/hero-list.component.html

```diff
@@ -1,3 +1,27 @@
+<div class="d-flex align-items-center mb-3">
+  <input
+    type="text"
+    class="form-control me-3"
+    style="max-width: 300px"
+    placeholder="Search heroes…"
+    [value]="filter()"
+    (input)="filter.set($any($event.target).value)"
```

Comment:
the $any is just as much bad practice as it was in the backend....
you can almost always avoid it, this case included.

```suggestion
  <input
    #searchHeroes
    type="text"
    class="form-control me-3"
    style="max-width: 300px"
    placeholder="Search heroes…"
    [value]="filter()"
    (input)="filter.set(searchHeroes.value)"
```

Reply:
same for the isRetired checkbox